### PR TITLE
Don't create default mongo clients when not needed and fix @MongoClientName qualifier

### LINKED ISSUE #11491: Named mongo client injection trying to use default mongo client
**Describe the bug**
When using the named mongo client injection, quarkus keeps trying to instantiate both the named client and the default mongo client. As a result, the default mongo client is getting the default connection string from `quarkus.mongodb.connection-string` which is always `localhost:27017`. Setting this property should not be needed, as it should not be used. Only the named mongo client connection string should be set in the application.properties file (e.g. quarkus.mongodb.fruits.connection-string)

**Expected behavior**
Setting only named mongo clients should only need the appropriate `quarkus.mongodb.XXX.connection-string` properties and `quarkus.mongodb.connection-string` must not be set at all. Moreover, when using the named mongo clients, the correct named connection must be used and NOT the default one (quarkus.mongodb.connection-string).

**Actual behavior**
When using only named mongo clients, quarkus is trying to instantiate the named mongo clients (as expected) AND the default one (not expected). Moreover, when for example an index is created upon @Startup with an injected named mongo client, quarkus picks the default client and not the named client as expected.

**To Reproduce**
1. clone https://github.com/agelbess/quarkus-mongodb-quickstart.git (contains the mongodb quickstart with a few changes)
2. fire up mongodb on 27017
3. change the properties of the application.properties file. Try to change the ports to valid and invali ones and see what happens in the logs
```
# default connection is always instantiated, why?
quarkus.mongodb.connection-string = mongodb://localhost:27017
# in FruitService, when creating the index, the default connection is used
quarkus.mongodb.fruits.connection-string = mongodb://localhost:27018
```

**Configuration**
```
# Configuration file
# default connection is always instantiated, why?
quarkus.mongodb.connection-string = mongodb://localhost:27017
# in FruitService, when creating the index, the default connection is used
quarkus.mongodb.fruits.connection-string = mongodb://localhost:27018
quarkus.mongodb.reactivefruits.connection-string = mongodb://localhost:27017
quarkus.mongodb.codecfruits.connection-string = mongodb://localhost:27017

```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`: 11
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.7
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): mvn

**Additional context**
This behaviour was not happening on quarkus v.1.4.

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #11491

### CODE REVIEW

**Comment by loicmathieu** [comment]
Is iterating over all the injection point has a cost ? I don't know how many of injection points exists in a classical application so I dont't know if it can add a lot of build time cost.

This will not work for MongoDB with Panache as we set all clients unremovable. 
Maybe it's not something that we should fix but instead providing a workaround by setting the default client `hosts`property to empty to disable the default MongoDB client:

```
# disable the default MongoDB client
quarkus.mongodb.hosts=
```

**Comment by geoand** [comment]
> Is iterating over all the injection point has a cost ? I don't know how many of injection points exists in a classical application so I dont't know if it can add a lot of build time cost.
> 

The cost is negligible. This same operation is performed in other extensions as well.

> This will not work for MongoDB with Panache as we set all clients unremovable.
> Maybe it's not something that we should fix but instead providing a workaround by setting the default client `hosts`property to empty to disable the default MongoDB client:
> 
> ```
> # disable the default MongoDB client
> quarkus.mongodb.hosts=
> ```

Yeah, as you said all will be made unremovable, but that should not be a problem.

**Comment by loicmathieu** [comment]
OK, so it fixes the issue only for direct usage of the MongoDB client not via Panache.

**Review by loicmathieu — APPROVED** [review]
LGTM

**Comment by geoand** [comment]
> OK, so it fixes the issue only for direct usage of the MongoDB client not via Panache.

It only fixes a part unfortunately. There is still the CDI related part that I mentioned in the issue that we need Martin to look at (I could of course, but it would take me a lot longer than Martin :))

**Review by mkouba — APPROVED** [review]
Looks good.

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java [hunk 1]
```diff
@@ -17,4 +17,10 @@
      */
     @ConfigItem(name = "metrics.enabled")
     public boolean metricsEnabled;
+
+    /**
+     * If set to true, the default clients will always be created even if there are no injection points that use them
+     */
+    @ConfigItem(name = "default-clients.force")
```
Comment (gsmet) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java:24]: A wise man said to me, don't use dots in config item names.

Also I don't see why `force-default-clients` would be a bad name.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some suggestion.

Also what worries me here is that we aren't consistent with the behavior elsewhere.

I think it all boils down to the "we need to be able to know if a runtime property will be set" discussion.

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java (hunk 1 above)
Comment (gsmet) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java:24]: Basically, the reason for no dots is that it doesn't work. I got hit by this in the Hibernate ORM config. No errors but it fails if you try to define the value in the application.properties.

If you really want a dot, use a group.

Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java:24]: Understood, I'll change it

**Comment by geoand** [comment]
> Added some suggestion.
> 
> Also what worries me here is that we aren't consistent with the behavior elsewhere.
> 
> I think it all boils down to the "we need to be able to know if a runtime property will be set" discussion.

Not really... If you know the injection points (which we do), then you know which mongo clients you need. Unless I am missing something?

**Comment by gsmet** [comment]
What if you programmatically get the bean? Isn't that a possibility?

**Comment by geoand** [comment]
This works for Panache and Camel because in those cases the bean does get created.

For users that would want to do that, then the new setting can be used.

**Comment by geoand** [comment]
@gsmet is this OK with you?

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 2]
```diff
@@ -52,12 +54,9 @@
 import io.quarkus.smallrye.health.deployment.spi.HealthBuildItem;
 
 public class MongoClientProcessor {
-    private static final DotName MONGOCLIENT_ANNOTATION = DotName.createSimple(MongoClientName.class.getName());
-
-    @BuildStep
-    BeanDefiningAnnotationBuildItem registerConnectionBean() {
-        return new BeanDefiningAnnotationBuildItem(MONGOCLIENT_ANNOTATION);
-    }
+    private static final DotName MONGO_CLIENT_ANNOTATION = DotName.createSimple(MongoClientName.class.getName());
```
Comment (gsmet) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:57]: If it didn't work before, please let's move this annotation to the API package i.e. no `runtime.`. It's user API.

Also we will probably need to update the doc and quickstart if it's mentioned.

**Review by gsmet — CHANGES_REQUESTED** [review]

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 2 above)
Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:57]: Agreed to do this in a follow up

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
This one has been backported via a specific PR due to conflicts: https://github.com/quarkusio/quarkus/pull/11849 .

**Pull request merged**